pkg/kubelet/images/pullmanager/image_pull_manager.go

```diff
@@ -0,0 +1,479 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pullmanager
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"slices"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/klog/v2"
+	kubeletconfiginternal "k8s.io/kubernetes/pkg/kubelet/apis/config"
+	kubecontainer "k8s.io/kubernetes/pkg/kubelet/container"
+	"k8s.io/kubernetes/pkg/util/parsers"
+)
+
+var _ ImagePullManager = &PullManager{}
+
+// writeRecordWhileMatchingLimit is a limit at which we stop writing yet-uncached
+// records that we found when we were checking if an image pull must be attempted.
+// This is to prevent unbounded writes in cases of high namespace turnover.
+const writeRecordWhileMatchingLimit = 100
+
+// PullManager is an implementation of the ImagePullManager. It
+// tracks images pulled by the kubelet by creating records about ongoing and
+// successful pulls.
+// It tracks the credentials used with each successful pull in order to be able
+// to distinguish tenants requesting access to an image that exists on the kubelet's
+// node.
+type PullManager struct {
+	recordsAccessor PullRecordsAccessor
+
+	requireCredentialVerification ImagePullPolicyEnforcer
+
+	imageService kubecontainer.ImageService
+
+	intentAccessors *namedLockSet  // image -> sync.Mutex
+	intentCounters  map[string]int // image -> number of current in-flight pulls
+
+	pulledAccessors *namedLockSet // imageRef -> sync.Mutex
+}
+
+func NewImagePullManager(ctx context.Context, recordsAccessor PullRecordsAccessor, imagePullPolicy ImagePullPolicyEnforcer, imageService kubecontainer.ImageService) (*PullManager, error) {
+	m := &PullManager{
+		recordsAccessor: recordsAccessor,
+
+		requireCredentialVerification: imagePullPolicy,
+
+		imageService: imageService,
+
+		intentAccessors: NewNamedLockSet(),
+		intentCounters:  make(map[string]int),
+
+		pulledAccessors: NewNamedLockSet(),
+	}
+
+	m.initialize(ctx)
+
+	return m, nil
+}
+
+func (f *PullManager) RecordPullIntent(image string) error {
+	f.intentAccessors.Lock(image)
+	defer f.intentAccessors.Unlock(image)
+
+	if err := f.recordsAccessor.WriteImagePullIntent(image); err != nil {
+		return fmt.Errorf("failed to record image pull intent: %w", err)
+	}
+
+	f.intentCounters[image]++
+	return nil
+}
+
+func (f *PullManager) RecordImagePulled(image, imageRef string, credentials *kubeletconfiginternal.ImagePullCredentials) {
+	if credentials == nil || (len(credentials.KubernetesSecrets) == 0 && !credentials.NodePodsAccessible) {
+		klog.ErrorS(
+			fmt.Errorf("credentials must not be nil, and they must contain at least one kubernetes secret coordinate or must set NodePodsAccessible=true"),
+			"attempted to write an ImagePulledRecord with invalid credentials",
+			"image", image, "imageRef", imageRef, "credentials", credentials)
+		return
```

Comment:
Can't we record an image pulled record with an empty CredentialMapping in that case, rather than erroring?  It seems like we could call `writePulledRecord(image, imageRef, credentials)` with nil or empty credentials

That would record that the kubelet was involved in pulling the image, but means any future attempts to use the image will have to reverify the credentials, right?

Reply:
I think I would be in favour of that rather than keeping the pull intent -> I'll make it so.